llmfoundry/models/hf/hf_base.py

```diff
@@ -0,0 +1,446 @@
+            use_cache=
+            False,  # Necessary due to https://github.com/huggingface/transformers/issues/28056
+        )
+
+        set_config_overrides(config, config_overrides)
+
+        return config
+
+    @classmethod
+    def build_metrics(
+        cls,
+        use_train_metrics: bool,
+        additional_train_metrics: Optional[list[str]] = None,
+        additional_eval_metrics: Optional[list[str]] = None,
+    ) -> tuple[list[Metric], list[Metric]]:
+        """Builds the training and evaluation metrics for the model.
+
+        Args:
+            use_train_metrics (bool): Whether to use training metrics.
+            additional_train_metrics (Optional[List[str]]): Additional training metrics to include.
+            additional_eval_metrics (Optional[List[str]]): Additional evaluation metrics to include.
+
+        Returns:
+            Tuple[List[Metric], List[Metric]]: A tuple containing the list of training metrics and evaluation metrics.
+        """
+        from llmfoundry.utils.builders import build_metric
+
+        train_metric_names = cls.default_train_metrics + (
+            additional_train_metrics or []
+        )
+        train_metrics = [
+            build_metric(metric, {}) for metric in train_metric_names
+        ] if use_train_metrics else []
+        eval_metric_names = cls.default_eval_metrics + (
+            additional_eval_metrics or []
+        )
+        eval_metrics = [
+            build_metric(metric, {}) for metric in eval_metric_names
+        ]
+
+        return train_metrics, eval_metrics
+
+    @classmethod
+    def build_inner_model(
+        cls,
+        pretrained_model_name_or_path: str,
+        pretrained_lora_id_or_path: Optional[str],
+        trust_remote_code: bool,
+        init_device: str,
+        use_flash_attention_2: bool,
+        use_auth_token: bool,
+        config_overrides: dict[str, Any],
+        load_in_8bit: bool,
+        pretrained: bool,
+        model_cls: Union[_BaseAutoModelClass,
+                         PreTrainedModel] = AutoModelForCausalLM,
+        prepare_for_fsdp: bool = False,
+    ) -> Union[PreTrainedModel, 'PeftModel']:
+        """Builds the inner model for the ComposerHFCausalLM.
+
+        Args:
+            pretrained_model_name_or_path (str): The pretrained model name or path.
+            pretrained_lora_id_or_path (Optional[str]): The pretrained LORA ID or path.
+            trust_remote_code (bool): Whether to trust remote code.
+            init_device (str): The initialization device.
+            use_flash_attention_2 (bool): Whether to use flash attention 2.
+            use_auth_token (bool): Whether to use an authentication token.
+            config_overrides (Dict[str, Any]): The configuration overrides.
+            load_in_8bit (bool): Whether to load in 8-bit.
+            pretrained (bool): Whether the model is pretrained.
+            model_cls (Union[Type, Type[PreTrainedModel]]): Kept for backwards compatibility.
```

Comment:
the current implementation doesn't really keep BC because it doesn't _use_ the `model_cls` arg. I would recommend defaulting it to `None`, and then, if it is not `None`, using it instead of the class attribute. That should preserve all functionality.

Reply:
added a line the sets `model_cls` the way you described. I did change the typing to be Optional for param. Hopefully that doesn't break BC

Reply:
yes, that is fine, because the new version is _more_ permissive.